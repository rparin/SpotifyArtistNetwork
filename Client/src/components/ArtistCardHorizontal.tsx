import React from "react";
import { ExternalLink } from "lucide-react";
import Image from "next/image";
import HorizontalList from "./UI/HorizontalList";
import { formatNumber, imageLoader } from "@/lib/utils";

export interface artistCardType extends React.HtmlHTMLAttributes<HTMLElement> {
  name: string;
  img: string;
  alt: string;
  genres: Array<string>;
  followers: string;
  pop: string;
  url: string;
  id: string;
}

const ArtistCardHorizontal = React.forwardRef<HTMLElement, artistCardType>(
  ({ id, name, img, alt, genres, followers, pop, url }, ref) => {
    return (
      <section ref={ref} className="relative w-96 rounded-md">
        <article className="flex items-center rounded-md py-2 gap-2 backdrop-blur-md px-2">
          <Image
            loader={() => {
              return imageLoader(img);
            }}
            className=" h-24 w-24 object-cover rounded-[50%]"
            width={100}
            height={100}
            src={img}
            alt={alt}
          />
          <div className="flex flex-col gap-1 w-64">
            <h2 className="flex text-xl font-medium line-clamp-1 justify-center text-center">
              {name}
            </h2>
            <div className="grid grid-cols-2 gap-1">
              <div className="flex gap-2 bg-slate-200/40 dark:bg-slate-600/60 py-1 justify-center">
                <h3 className="text-xs font-normal">Followers</h3>
                <p className="font-medium text-xs">
                  {formatNumber(parseInt(followers))}
                </p>
              </div>
              <div className="flex gap-2 bg-slate-200/40 dark:bg-slate-600/60 py-1 justify-center">
                <h3 className="text-xs font-normal">Popularity</h3>
                <p className="font-medium text-xs">{pop}</p>
              </div>
            </div>
            {genres.length > 0 && (
              <HorizontalList
                id={id}
                items={genres}
                className="dark:bg-slate-500/80 bg-slate-200/40 text-xs"
              />
            )}

            <a
              href={url}
              target="_blank"
              rel="noopener noreferrer"
              className="flex gap-1 text-xs justify-center items-center text-center w-full rounded-br-md bg-[#1DB954] horizontal-mask">
              Spotify
              <ExternalLink className="w-auto h-[.7rem] mb-2" />
            </a>
          </div>
        </article>
        <Image
          loader={() => {
            return imageLoader(img);
          }}
          className="absolute top-0 left-0 -z-10 h-full w-full opacity-80 dark:opacity-40 rounded-md"
          width={100}
          height={100}
          src={img}
          alt={alt}
        />
      </section>
    );
  }
);

ArtistCardHorizontal.displayName = "ArtistCardHorizontal";

export { ArtistCardHorizontal };
