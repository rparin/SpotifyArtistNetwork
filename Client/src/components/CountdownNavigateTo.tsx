"use client";
import { useRouter } from "next/navigation";
import { useCountdown } from "@/hooks/useCountdown";

export default function CountdownNavigateTo(props: {
  href: string;
  countdown: number;
}) {
  const router = useRouter();
  const ct = useCountdown(props.countdown, () => {
    router.push(props.href);
  });
  return <span>{ct / 1000}</span>;
}
