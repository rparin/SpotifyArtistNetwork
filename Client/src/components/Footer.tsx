import { FOOTER_PRIVACY, FOOTER_SEARCH_TEXT, CR } from "@/constants";
import { ExternalLink } from "lucide-react";
import { cn } from "@/lib/utils";

export default function Footer(props: { className?: string }) {
  return (
    <div
      className={cn(
        "bg-black text-white mx-auto flex flex-col items-center gap-2 py-2 w-full z-50",
        props.className
      )}>
      {/* <section className="flex gap-5 text-sm">
        <a
          href=""
          target="_blank"
          rel="noopener noreferrer"
          className="hover:underline underline-offset-2">
          <p>{FOOTER_SEARCH_TEXT}</p>
        </a>
        <a
          href=""
          target="_blank"
          rel="noopener noreferrer"
          className="hover:underline underline-offset-2">
          <p>{FOOTER_PRIVACY}</p>
        </a>
      </section> */}
      <section className="flex">
        <p className="text-white/30 text-xs">{CR}</p>
        <a
          href="https://rparin.github.io"
          aria-label="External link to rparin portfolio website"
          target="_blank"
          rel="noopener noreferrer">
          <ExternalLink className="opacity-40 w-auto h-[.7rem]" />
        </a>
      </section>
    </div>
  );
}
