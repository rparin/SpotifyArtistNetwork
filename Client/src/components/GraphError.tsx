export default function GraphError(props: { errorMsg: string }) {
  return (
    <div className="flex h-screen justify-center text-center items-center">
      <h2 className="flex flex-col text-base">
        <span className="text-lg font-semibold">{props.errorMsg}</span>
        <span className="font-normal">Please try again at a later time</span>
      </h2>
    </div>
  );
}
