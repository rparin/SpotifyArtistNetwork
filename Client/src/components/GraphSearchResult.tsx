import Image from "next/image";

export default function GraphSearchResult(props: any) {
  return (
    <>
      <div className="flex gap-3">
        <Image
          className="h-7 w-7 object-cover rounded-[50%]"
          width={100}
          height={100}
          src={props.img}
          alt={`${props.name} Spotify profile pic`}
        />
        {props.name}
      </div>
    </>
  );
}
