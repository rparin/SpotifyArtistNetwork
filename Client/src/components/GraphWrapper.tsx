import dynamic from "next/dynamic";

const ForceGraph = dynamic(() => import("../components/Graph"), {
  ssr: false,
});

export default ForceGraph;
