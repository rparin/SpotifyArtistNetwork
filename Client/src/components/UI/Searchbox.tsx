"use client";
import React, { useState, useRef, useEffect } from "react";
import { cn } from "@/lib/utils";
import { cva, type VariantProps } from "class-variance-authority";
import { Search } from "lucide-react";
import { useRouter } from "next/navigation";

const searchVariants = cva(
  "rounded-[7rem] w-full px-4 text-sm md:text-base shadow-lg focus:shadow-xl dark:bg-background",
  {
    variants: {
      variant: {
        default: "outline outline-black outline-2 dark:outline-white",
        secondary: "bg-black dark:bg-white text-white dark:text-black",
      },
    },
    defaultVariants: {
      variant: "default",
    },
  }
);

const sizeVariants = cva("relative h-10 flex justify-start", {
  variants: {
    size: {
      default: "w-5/6 md:w-9/12 lg:w-4/12",
      sm: "w-[60%] md:w-[50%] lg:w-[30%]",
      md: "w-[80%] md:w-[50%] lg:w-[40%]",
    },
  },
  defaultVariants: {
    size: "default",
  },
});

const searchIconVariants = cva(
  "absolute rounded-md z-50 top-1 right-3 w-auto h-8 p-1",
  {
    variants: {
      variant: {
        default: "bg-white dark:bg-background",
        secondary: "invert bg-white dark:bg-black",
      },
    },
    defaultVariants: {
      variant: "default",
    },
  }
);

export interface SearchProps
  extends React.HtmlHTMLAttributes<HTMLInputElement>,
    VariantProps<typeof searchVariants>,
    VariantProps<typeof sizeVariants> {
  placeholder?: string;
  inputOnLoad?: string | undefined | null;
  route?: string;
}

const Searchbox = ({
  className,
  variant,
  size,
  placeholder,
  inputOnLoad,
  route = "/search",
  ...props
}: SearchProps) => {
  // Hold search values
  const [searchValue, setSearch] = useState("");
  const [prevValue, setPrev] = useState("");

  //Values for page load
  const ref = useRef<HTMLInputElement>(null);

  // Values for search use
  const router = useRouter();

  // Input value into searchbox on page load
  useEffect(() => {
    if (ref.current && inputOnLoad) {
      ref.current.value = inputOnLoad;
    }

    if (ref.current) {
      ref.current.focus();
    }
  }, []);

  // Set search value on input
  const handleInputChange = (event: {
    target: EventTarget & HTMLInputElement;
  }) => {
    setSearch(event.target.value);
  };

  // Set query param and go to search page
  const handleClick = () => {
    if (searchValue == "" || prevValue == searchValue) return;
    router.push(`${route}?q=${searchValue}`);
    setPrev(searchValue);
  };

  const handleKeyDown = (event: { key: string }) => {
    if (event.key === "Enter") {
      handleClick();
    }
  };

  return (
    <div className={cn(sizeVariants({ size }))}>
      <input
        ref={ref}
        className={cn(searchVariants({ variant }), className)}
        type="text"
        placeholder={placeholder}
        onKeyDown={handleKeyDown}
        onChange={handleInputChange}
        {...props}
      />

      <button onClick={handleClick}>
        <Search className={searchIconVariants({ variant })} />
      </button>
    </div>
  );
};

export { Searchbox };
