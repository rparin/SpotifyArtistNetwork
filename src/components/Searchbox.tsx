"use client";
import React, { useState, useRef, useEffect } from "react";
import { cva, type VariantProps } from "class-variance-authority";
import { Search, X } from "lucide-react";
import { useRouter } from "next/navigation";
import cn from "@/utils/cn";

const searchVariants = cva(
  "rounded-[7rem] w-full px-5 text-sm md:text-base shadow-lg focus:shadow-xl dark:bg-background",
  {
    variants: {
      variant: {
        default: "outline outline-black outline-2 dark:outline-white",
        secondary: "bg-black dark:bg-white text-white dark:text-black",
      },
    },
    defaultVariants: {
      variant: "default",
    },
  }
);

const sizeVariants = cva("relative h-10 flex justify-start", {
  variants: {
    size: {
      default: "w-[80%] md:w-[60%] max-w-2xl",
      sm: "w-[60%] md:w-[40%] lg:w-[30%]",
      md: "w-[80%] md:w-[50%] lg:w-[40%]",
    },
  },
  defaultVariants: {
    size: "default",
  },
});

const searchIconVariants = cva(
  "absolute rounded-md z-50 top-1 right-3 w-auto h-8 p-1",
  {
    variants: {
      variant: {
        default: "bg-white dark:bg-background",
        secondary: "invert bg-white dark:bg-black",
      },
    },
    defaultVariants: {
      variant: "default",
    },
  }
);

export interface SearchProps
  extends React.HtmlHTMLAttributes<HTMLInputElement>,
    VariantProps<typeof searchVariants>,
    VariantProps<typeof sizeVariants> {
  placeholder?: string;
  inputOnLoad?: string | undefined | null;
  route?: string;
  queryRoute?: boolean;
  focus?: boolean;
}

const Searchbox = ({
  className,
  variant,
  size,
  placeholder,
  inputOnLoad,
  route = "/search",
  queryRoute = true,
  focus = false,
  ...props
}: SearchProps) => {
  // Hold search values
  const [searchValue, setSearch] = useState("");
  const [prevValue, setPrev] = useState("");

  const setHelper = (setFunc: (arg0: string) => void, val: string) => {
    setFunc(val.toLowerCase());
  };

  //Values for page load
  const ref = useRef<HTMLInputElement>(null);

  // Values for search use
  const router = useRouter();

  const changeInput = (val: string, prev: string = searchValue) => {
    if (ref.current) {
      ref.current.value = val;
      if (focus) ref.current.focus();
    }
    setHelper(setPrev, prev);
    setHelper(setSearch, val);
  };

  // Input value into searchbox on page load
  useEffect(() => {
    if (inputOnLoad && ref.current) {
      ref.current.value = inputOnLoad;
      setHelper(setPrev, inputOnLoad);
      setHelper(setSearch, inputOnLoad);
    }
    if (ref.current) {
      ref.current.focus();
    }
  }, [inputOnLoad]);

  // Set search value on input
  const handleInputChange = (event: {
    target: EventTarget & HTMLInputElement;
  }) => {
    setHelper(setSearch, event.target.value);
  };

  // Go to search page
  const handleClick = () => {
    if (searchValue == "" || prevValue == searchValue) {
      if (ref.current) {
        ref.current.focus();
      }
      return;
    }
    if (queryRoute) {
      router.push(`${route}?q=${searchValue}`);
    } else {
      router.push(`${route}/${searchValue}`);
    }
    setHelper(setPrev, searchValue);
  };

  const handleKeyDown = (event: { key: string }) => {
    if (event.key === "Enter") {
      handleClick();
    }
  };

  const getButton = () => {
    if (prevValue == searchValue && searchValue != "") {
      return (
        <button title="Search button" onClick={() => changeInput("")}>
          <X className={searchIconVariants({ variant })} />
        </button>
      );
    }
    return (
      <button title="Search button" onClick={handleClick}>
        <Search className={searchIconVariants({ variant })} />
      </button>
    );
  };

  const handleOnSubmit = (e: React.FormEvent) => {
    e.preventDefault();
    handleClick();
  };

  return (
    <form
      role="search"
      onSubmit={handleOnSubmit}
      className={cn(sizeVariants({ size }))}>
      <label
        htmlFor="searchbar"
        aria-label="Spotify Artist Network Search Bar"></label>
      <input
        name="searchbar"
        id="searchbar"
        className={cn(searchVariants({ variant }), className)}
        ref={ref}
        type="text"
        placeholder={placeholder}
        onChange={handleInputChange}
        {...props}
      />
      {getButton()}
    </form>
  );
};

export { Searchbox };
