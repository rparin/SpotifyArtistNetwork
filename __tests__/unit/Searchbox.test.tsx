import { render } from "@testing-library/react";
import { axe, toHaveNoViolations } from "jest-axe";
import { Searchbox } from "@/components/Searchbox";

test("it renders component unchanged", () => {
  const { container } = render(<Searchbox />);
  expect(container).toMatchSnapshot();
});

describe("testing Footer Accessibility", () => {
  expect.extend(toHaveNoViolations);

  test("it should have no accessibility violations", async () => {
    const { container } = render(<Searchbox />);
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });
});
