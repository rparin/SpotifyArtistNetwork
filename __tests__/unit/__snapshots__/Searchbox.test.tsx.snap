// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders component unchanged 1`] = `
<div>
  <form
    class="relative h-10 flex justify-start w-[80%] md:w-[60%] max-w-2xl"
    role="search"
  >
    <label
      aria-label="Spotify Artist Network Search Bar"
      for="searchbar"
    />
    <input
      class="rounded-[7rem] w-full px-5 text-sm md:text-base shadow-lg focus:shadow-xl dark:bg-background outline outline-black outline-2 dark:outline-white"
      id="searchbar"
      name="searchbar"
      type="text"
    />
    <button
      title="Search button"
    >
      <svg
        class="lucide lucide-search absolute rounded-md z-50 top-1 right-3 w-auto h-8 p-1 bg-white dark:bg-background"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <circle
          cx="11"
          cy="11"
          r="8"
        />
        <path
          d="m21 21-4.3-4.3"
        />
      </svg>
    </button>
  </form>
</div>
`;
